opencog/atomspace/AtomTable.cc

```diff
@@ -498,23 +495,45 @@ Handle AtomTable::add(AtomPtr atom, bool async)
     atom->keep_incoming_set();
     atom->setAtomSpace(_as);
 
+    ContentHash ch = atom->get_hash();
+
+    // Lock, to prevent two different threads from trying to add exactly
+    // the same atom.
+    std::unique_lock<std::recursive_mutex> lck(_store_mtx);
```

Comment:
well, this lock has to include the state-link check, as otherwise the statelink will get garbled.

I do not beleive we have a multi-threaded safety test for the statelink, just right now.  I'd have to think about that,